pkg/kv/kvclient/kvcoord/dist_sender.go

```diff
@@ -2557,71 +2543,83 @@ func (ds *DistSender) sendToReplicas(
+// cross-region and cross-zone.
+func (ds *DistSender) getCrossLocalityComparison(
 	ctx context.Context, ba *kvpb.BatchRequest,
-) (bool, bool) {
+) roachpb.LocalityComparisonType {
 	gatewayNodeDesc, err := ds.nodeDescs.GetNodeDescriptor(ba.GatewayNodeID)
 	if err != nil {
 		log.VEventf(ctx, 2, "failed to perform look up for node descriptor %s", err)
-		return false, false
+		return roachpb.LocalityComparisonType_UNDEFINED
 	}
 	destinationNodeDesc, err := ds.nodeDescs.GetNodeDescriptor(ba.Replica.NodeID)
 	if err != nil {
 		log.VEventf(ctx, 2, "failed to perform look up for node descriptor %s", err)
-		return false, false
+		return roachpb.LocalityComparisonType_UNDEFINED
 	}
-	isCrossRegion, regionErr, isCrossZone, zoneErr := gatewayNodeDesc.Locality.IsCrossRegionCrossZone(destinationNodeDesc.Locality)
+
+	comparisonResult, regionErr, zoneErr := gatewayNodeDesc.Locality.CompareWithLocality(destinationNodeDesc.Locality)
 	if regionErr != nil {
-		log.VEventf(ctx, 2, "%v", regionErr)
+		log.VEventf(ctx, 2, "unable to determine if batch is cross region %v", regionErr)
 	}
 	if zoneErr != nil {
-		log.VEventf(ctx, 2, "%v", zoneErr)
+		log.VEventf(ctx, 2, "unable to determine if batch is cross zone %v", zoneErr)
 	}
-	return isCrossRegion, isCrossZone
+	return comparisonResult
 }
 
-// checkAndUpdateBatchRequestMetrics updates the batch requests metrics in a
-// more meaningful way. Cross-region metrics monitor activities across different
-// regions. Cross-zone metrics monitor cross-zone activities within the same
-// region or in cases where region tiers are not configured. The check result is
-// returned here to avoid redundant check for metrics updates after receiving
-// batch responses.
-func (ds *DistSender) checkAndUpdateBatchRequestMetrics(
+// checkAndUpdateCrossLocalityBatchMetrics updates the batch requests metrics in
+// a more meaningful way. Cross-region metrics monitor activities across
+// different regions. Cross-zone metrics monitor cross-zone activities within
+// the same region or in cases where region tiers are not configured. The
+// locality comparison result is returned here to avoid redundant check for
+// metrics updates after receiving batch responses.
+func (ds *DistSender) checkAndUpdateCrossLocalityBatchMetrics(
 	ctx context.Context, ba *kvpb.BatchRequest,
-) (shouldIncCrossRegion bool, shouldIncCrossZone bool) {
-	ds.metrics.ReplicaAddressedBatchRequestBytes.Inc(int64(ba.Size()))
-	isCrossRegion, isCrossZone := ds.isCrossRegionCrossZoneBatch(ctx, ba)
-	if isCrossRegion {
-		if !isCrossZone {
-			log.VEventf(ctx, 2, "unexpected: cross region but same zone")
-		} else {
-			ds.metrics.CrossRegionBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossRegion = true
-		}
-	} else {
-		if isCrossZone {
-			ds.metrics.CrossZoneBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossZone = true
-		}
+) roachpb.LocalityComparisonType {
+	if ds.TestingBatchInterceptor != nil {
```

Comment:
Splitting the second commit into another [PR](https://github.com/cockroachdb/cockroach/pull/105267/files). 

I agree that using a testing knob and the filter strategy wasn’t ideal. It was challenging to set up the tests since we wanted some batches to go through for test setup while skipping the metric updates. I’ve also tried blocking mechanisms to try to measure the difference before and after metrics right after one single batch goes through. It was quite complicated since there were lots of concurrent updates. In the end, it was working but required lots of interceptors. And it felt like it wasn’t testing what we wanted to test. Using the filter on the metrics update was the alternative. 

I refactored these tests and adopted unit tests for DistSender and Node metrics updates.